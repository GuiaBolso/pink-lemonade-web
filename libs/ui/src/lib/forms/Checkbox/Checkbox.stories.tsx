import React from 'react';
import Checkbox from './Checkbox';

export default {
  title: 'Forms/Checkbox',
  component: Checkbox,
};

export const simple = () => (
  <Checkbox label={<span>Receber newsletter</span>} name="newsletter" />
);

export const standAlone = () => (
  <Checkbox
    label={<span>Receber newsletter</span>}
    variant="stand-alone"
    name="newsletter"
  />
);

export const checked = () => (
  <Checkbox
    defaultChecked
    label={<span>Receber newsletter</span>}
    name="newsletter"
  />
);

export const disabled = () => (
  <Checkbox
    label={<span>Receber newsletter</span>}
    checked
    disabled
    name="newsletter"
  />
);
