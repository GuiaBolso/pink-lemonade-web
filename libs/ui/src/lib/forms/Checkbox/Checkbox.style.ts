import styled from '@emotion/styled';
import { css, Theme } from '@emotion/react';

import MuiCheckbox from '@material-ui/core/Checkbox';
import MuiFormControlLabel from '@material-ui/core/FormControlLabel';

type CheckboxStyledProps = {
  theme?: Theme;
  variant: 'simple' | 'stand-alone';
};

export const FormControlLabel = styled(MuiFormControlLabel)`
  ${({ theme, variant }: CheckboxStyledProps) => {
    const colors = {
      simple: theme?.colors?.neutral?.dark,
      'stand-alone': theme?.colors?.brand?.secondary?.default,
    };

    return css`
      && {
        .MuiCheckbox-root {
          color: ${colors[variant] || colors.simple};
          &.Mui-checked {
            color: ${theme?.colors?.brand?.secondary?.default};
          }
          &.Mui-disabled {
            color: ${theme?.colors?.neutral?.light};
          }

          &:hover {
            background-color: ${theme?.colors?.brand?.secondary?.lightest};
          }
        }
      }

      .MuiTypography-root {
        font-size: inherit;
        font-family: ${theme?.type?.fontFamily?.neutral};
        font-weight: inherit;
        line-height: inherit;
        letter-spacing: inherit;

        &.Mui-disabled {
          color: ${theme?.colors?.neutral?.light};
        }
      }
    `;
  }};
`;

export const Checkbox = styled(MuiCheckbox)``;
