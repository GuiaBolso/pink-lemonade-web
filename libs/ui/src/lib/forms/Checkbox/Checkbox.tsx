import React, { memo } from 'react';
import {
  FormControlLabelProps,
  CheckboxProps as MuiCheckboxProps,
} from '@material-ui/core';

import * as S from './Checkbox.style';

type CheckboxProps = {
  name?: string;
  variant?: 'simple' | 'stand-alone';
} & Partial<FormControlLabelProps>;

const Checkbox = ({
  label,
  name,
  disabled = false,
  variant = 'simple',
  defaultChecked,
  ...rest
}: CheckboxProps) => {
  let checkboxProps: MuiCheckboxProps = {
    name,
  };

  defaultChecked &&
    (checkboxProps = {
      ...checkboxProps,
      defaultChecked,
    });

  return (
    <S.FormControlLabel
      {...rest}
      label={label}
      disabled={disabled}
      variant={variant}
      control={<S.Checkbox {...checkboxProps} />}
    />
  );
};

export default memo(Checkbox);
