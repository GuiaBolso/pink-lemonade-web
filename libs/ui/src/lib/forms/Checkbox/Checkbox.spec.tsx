import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import Checkbox from './Checkbox';

describe('Checkbox Component', () => {
  const makeSetup = () => {
    const element = (
      <Checkbox label={<span>Lorem Ipsum</span>} name="newsletter" />
    );
    const utils = render(element);
    const input = utils.container.querySelector('input');
    const label = utils.container.querySelector(
      'span.MuiFormControlLabel-label',
    );

    return {
      element,
      input,
      label,
      ...utils,
    };
  };

  const setup = makeSetup();

  it('It should be same value in fired event', () => {
    const { input } = setup;
    fireEvent.change(input, { target: { checked: true } });
    expect(input.checked).toBe(true);
  });

  it('It should be correct label', () => {
    const { label } = setup;
    expect(label.textContent).toBe('Lorem Ipsum');
  });

  it('renders correctly', () => {
    const { element } = setup;
    const tree = renderer.create(element).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
