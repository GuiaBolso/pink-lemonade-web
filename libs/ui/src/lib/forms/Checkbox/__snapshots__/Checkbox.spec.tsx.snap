// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component renders correctly 1`] = `
<label
  className="MuiFormControlLabel-root css-1l8qi0x"
  variant="simple"
>
  <span
    aria-disabled={false}
    className="MuiButtonBase-root MuiIconButton-root jss5 MuiCheckbox-root MuiCheckbox-colorSecondary css-0 MuiIconButton-colorSecondary"
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={null}
  >
    <span
      className="MuiIconButton-label"
    >
      <input
        className="jss8"
        data-indeterminate={false}
        disabled={false}
        name="newsletter"
        onChange={[Function]}
        type="checkbox"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
  </span>
  <span
    className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
  >
    <span>
      Lorem Ipsum
    </span>
  </span>
</label>
`;
